crypto/shamir.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Implementation of the hazardous parts of the SSS library
+ *
+ * Copyright (c) 2017 Daan Sprenkels <hello@dsprenkels.com>
+ * Copyright (c) 2019 SatoshiLabs
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * This code contains the actual Shamir secret sharing functionality. The
+ * implementation of this code is based on the idea that the user likes to
+ * generate/combine 32 shares (in GF(2^8)) at the same time, because a 256 bit
+ * key will be exactly 32 bytes. Therefore we bitslice all the input and
+ * unbitslice the output right before returning.
+ *
+ * This bitslice approach optimizes natively on all architectures that are 32
+ * bit or more. Care is taken to use not too many registers, to ensure that no
+ * values have to be leaked to the stack.
+ *
+ * All functions in this module are implemented constant time and constant
+ * lookup operations, as all proper crypto code must be.
+ */
+
+#include <string.h>
+#include "memzero.h"
+#include "shamir.h"
+
+static void bitslice(uint32_t r[8], const uint8_t *x, size_t len) {
+  size_t bit_idx, arr_idx;
+  uint32_t cur;
+
+  memset(r, 0, sizeof(uint32_t[8]));
+  for (arr_idx = 0; arr_idx < len; arr_idx++) {
+    cur = (uint32_t)x[arr_idx];
+    for (bit_idx = 0; bit_idx < 8; bit_idx++) {
+      r[bit_idx] |= ((cur & (1 << bit_idx)) >> bit_idx) << arr_idx;
+    }
+  }
+}
+
+static void unbitslice(uint8_t *r, const uint32_t x[8], size_t len) {
+  size_t bit_idx, arr_idx;
+  uint32_t cur;
+
+  memset(r, 0, sizeof(uint8_t) * len);
+  for (bit_idx = 0; bit_idx < 8; bit_idx++) {
+    cur = (uint32_t)x[bit_idx];
+    for (arr_idx = 0; arr_idx < len; arr_idx++) {
+      r[arr_idx] |= ((cur & (1 << arr_idx)) >> arr_idx) << bit_idx;
+    }
+  }
+}
+
+static void bitslice_setall(uint32_t r[8], const uint8_t x) {
+  size_t idx;
+  for (idx = 0; idx < 8; idx++) {
+    r[idx] = ((int32_t)((x & (1 << idx)) << (31 - idx))) >> 31;
```

Comment:
The resulting value of a right shift of a negative number is compiler dependent.
```suggestion
    r[idx] = -((x & (1 << idx)) >> idx);
```

Reply:
How about `-((x >> idx) & 1);`?

Reply:
Even better.

Reply:
Resolved by https://github.com/trezor/trezor-firmware/pull/85/commits/223868f552404c9f1df72691ead9581beec169d0.